[Docs] Update EC2 Discovery guide for bootstrapping. (#18308)
gravitational/teleport: docs/pages/server-access/guides/ec2-discovery.mdx

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM
policies.
- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the
default Teleport install script. (For other Linux distributions, you can
install Teleport manually.)

(!docs/pages/includes/tctl.mdx!)

## Step 1/6. Create an EC2 invite token

When discovering EC2 instances, Teleport makes use of IAM invite tokens for

> authenticating joining Nodes. Read [Joining Nodes via AWS IAM Role](../../management/guides/joining-nodes-aws-iam.mdx)

Reviewer: What additional information should the reader expect to get out of the linked guide? If it's further reading for general edification, I think the link should go in "Next steps" instead of this section.

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- AWS account with EC2 instances and permissions to create and attach IAM

> policies as well as permission to create Systems Manager documents.

Author: ```suggestion
policies.
```

I'm not sure if the user's AWS account also needs this permission any more with the new bootstrap command. Defer to @lxea.

- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the
default Teleport install script. (For other Linux distributions, you can
install Teleport manually.)

(!docs/pages/includes/tctl.mdx!)

## Step 1/5. Define an IAM policy

The `teleport discovery bootstrap` command will automate the process of
defining and implementing IAM policies required to make auto-discovery work. It
requires only a single pre-defined policy, attached to the EC2 instance running
the command:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [

>                 "ssm:SendCommand",
>                 "ec2:DescribeInstances",
>                 "ssm:GetCommandInvocation",

Author: ```suggestion
```

>                 "ssm:GetCommandInvocation",

Author: Removed

                "iam:GetPolicy",
                "iam:TagPolicy",
                "iam:ListPolicyVersions",
                "iam:CreatePolicyVersion",
                "iam:CreatePolicy",
                "ssm:CreateDocument",
                "iam:DeletePolicyVersion",
                "iam:AttachRolePolicy",
                "iam:PutRolePermissionsBoundary"
            ],
            "Resource": "*"
        }
    ]
}
```

Create this policy and apply it to the Node (EC2 instance) that will run the Discovery Service.

> ## Step 2/5. Install Teleport on the Discovery Node

Author: Re-added.

<Admonition type="tip">

If you plan on running the Discovery Service on the same node already running
another Teleport service (Auth or Proxy, for example), you can skip this step.

</Admonition>

Install Teleport on the EC2 instance that will run the Discovery Service:

(!docs/pages/includes/install-linux.mdx!)

## Step 3/5. Configure Teleport to discover EC2 instances

Assuming a discrete Node for the Discovery Service, the service requires a
valid auth token to connect to the cluster. Generate one by running the
following command against your Teleport Auth Service:

```code
$ tctl tokens add --type=discovery
```

Save the generated token in `/tmp/token` on the Node (EC2 instance) that will
run the Discovery Service:

In order to enable EC2 instance discovery the `discovery_service.aws` section

> of `teleport.yaml` must include at least one entry:

Author: @ptgott seems like a good use case for those fancy new vars, maybe?

(!docs/pages/includes/tctl.mdx!)

## Step 1/6. Create an EC2 invite token

> When discovering EC2 instances, Teleport makes use of IAM invite tokens for

Author: Amended in https://github.com/gravitational/teleport/pull/18308/commits/1762ae809414776cd68cf3db508f27bb96b8ab89

> When discovering EC2 instances, Teleport makes use of IAM invite tokens for

Reviewer: Until this point, there's nothing to indicate how EC2 Discovery works and why we need an invite token. I think a sentence about how these tokens work within a Teleport cluster would go a long way toward explaining, say, why we would want to set a longer TTL on the token. For example,  does the SSH Service poll the EC2 APIs in the configured regions until the token TTL elapses? 

We also don't mention SSM until toward the end of the guide, so it's not really clear architecturally speaking what EC2 Auto-Discovery looks like unless the reader figures it out on their own as they go through the guide.

Also, per the [style guide](https://goteleport.com/docs/contributing/documentation/style-guide/#getting-started-articles), we're trying to avoid external links in tutorials. If we're trying to get the user through the tutorial as quickly as possible so they can see value from Teleport, telling them to read another docs page without a clear goal for doing so adds time to complete the guide and risks distracting the user because it requires shifting context.

> When discovering EC2 instances, Teleport makes use of IAM invite tokens for

Reviewer: That sounds good to me!

authenticating joining Nodes.

Create a file called `token.yaml`:

```yaml
# token.yaml
kind: token
version: v2
metadata:
  # the token name is not a secret because instances must prove that they are
  # running in your AWS account to use this token
  name: aws-discovery-iam-token
  # set a long expiry time, as the default for tokens is only 30 minutes

>   expires: "3000-01-01T00:00:00Z"

Author: I would also ask @lxea to address this since this came from their issue.

>   expires: "3000-01-01T00:00:00Z"

Author: Sounds good to me, no @ptgott ?

>   expires: "3000-01-01T00:00:00Z"

Reviewer: Works for me! I had this question because I didn't understand how EC2 discovery works, which touches on earlier comment—it would be really nice to have an architectural overview of EC2 discovery at the beginning of the guide, because otherwise the way this feature works is pretty murky.

the command:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iam:GetPolicy",
                "iam:TagPolicy",
                "iam:ListPolicyVersions",
                "iam:CreatePolicyVersion",
                "iam:CreatePolicy",
                "ssm:CreateDocument",
                "iam:DeletePolicyVersion",
                "iam:AttachRolePolicy",
                "iam:PutRolePermissionsBoundary"
            ],
            "Resource": "*"
        }
    ]
}
```

Create this policy and apply it to the Node (EC2 instance) that will run the Discovery Service.

## Step 3/6. Install Teleport on the Discovery Node

<Admonition type="tip">

If you plan on running the Discovery Service on the same Node already running
another Teleport service (Auth or Proxy, for example), you can skip this step.

</Admonition>

Install Teleport on the EC2 instance that will run the Discovery Service:

(!docs/pages/includes/install-linux.mdx!)

## Step 4/6. Configure Teleport to discover EC2 instances

If you are running the Discovery Service on its own host, the service requires a

> valid auth token to connect to the cluster. Generate one by running the

Reviewer: ```suggestion
valid invite token to connect to the cluster. Generate one by running the
```
So we have consistent language around these tokens

       "env": "prod" # Match EC2 instances where tag:env=prod
```

> - Edit the `teleport.auth_servers` key to match your Auth or Proxy server URI

Reviewer: ```suggestion
Edit the `teleport.auth_servers` key to match your Auth Service or Proxy Service's URI
```

Author: I'm 👍 on the changes at the end of the sentence, but why remove the UL marker?

Reviewer: That was a mistake—I'll fix the suggestion

  and port.
- Adjust the keys under `discovery_service.aws` to match your EC2 environment,

>   specifically the regions and tags you want to associate with the discovery

Reviewer: ```suggestion
  specifically the regions and tags you want to associate with the Discovery
```

>   service.

Reviewer: ```suggestion
  Service.
```

    echo Teleport-{{ .MajorVersion }}
version: v1
```

Which, when retrieved by the Systems Manager document will evaluate to a script
with the following contents:

```sh
echo teleport.example.com
echo Teleport-(=teleport.version=)
```

The default installer will take the following actions:

- Add an official Teleport repository to supported Linux distributions.
- Install Teleport via `apt` or `yum`.
- Generate the Teleport config file and write it to `/etc/teleport.yaml`.
- Enable and start the Teleport service.

## Troubleshooting

If Installs are showing failed or instances are failing to appear check the
Command history in AWS System Manager -> Node Management -> Run Command.
Select the instance-id of the Target to review Errors.

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```
While we're editing this guide, I've suggested making this heading sentence case per the style guide.

Information on IAM best practices on EC2 instances managed by Systems

> Manager can be found for IAM best practices on EC2 instances managed

Reviewer: I think there's an unintended repetition of "EC2 instances managed by Systems Manager" here
